src/h3/mod.rs

```diff
@@ -26,4 +26,830 @@
+                            hdr.name().as_bytes(),
+                        )
+                        .unwrap(),
+                        http::header::HeaderValue::from_str(hdr.value()).unwrap(),
+                    );
+                },
+            }
+        }
+
+        *resp.status_mut() = StatusCode::from_bytes(status.as_bytes()).unwrap();
+        *resp.version_mut() = http::Version::HTTP_2;
+
+        // debug!("Prepared response {:?}", resp);
+
+        resp
+    }
+}
+
+impl H3Connection {
+    #[allow(clippy::new_ret_no_self)]
+    fn new(config: &mut H3Config, is_server: bool) -> Result<Box<H3Connection>> {
+        let root = String::clone(&config.root_dir); // TODO shouldn't need to clone here
+
+        Ok(Box::new(H3Connection {
+            root_dir: root,
+            is_server,
+            num_placeholders: config.num_placeholders,
+            max_header_list_size: config.max_header_list_size,
+            qpack_max_table_capacity: config.qpack_max_table_capacity,
+            qpack_blocked_streams: config.qpack_blocked_streams,
+
+            peer_num_placeholders: None,
+            peer_max_header_list_size: None,
+            peer_qpack_max_table_capacity: None,
+            peer_qpack_blocked_streams: None,
+
+            control_stream_open: false,
+            peer_control_stream_open: false,
+
+            qpack_encoder: qpack::Encoder::new(),
+            qpack_decoder: qpack::Decoder::new(),
+
+            qpack_encoder_stream_open: false,
+            peer_qpack_encoder_stream_open: false,
+            qpack_decoder_stream_open: false,
+            peer_qpack_decoder_stream_open: false,
+        }))
+    }
+
+    fn get_control_stream_id(&mut self) -> u64 {
+        // TODO get an available unidirectional stream ID more nicely
+        if self.is_server {
+            return 0x3;
+        } else {
+            return 0x2;
+        }
+    }
+
+    fn is_control_stream(stream_id: u64) -> bool {
+        // TODO make this respect stream types and NOT assume types from ID
+        return stream_id == 0x2 || stream_id == 0x3;
+    }
+
+    fn get_encoder_stream_id(&mut self) -> u64 {
+        // TODO get an available unidirectional stream ID more nicely
+        if self.is_server {
+            return 0x7;
+        } else {
+            return 0x6;
+        }
+    }
+
+    fn get_decoder_stream_id(&mut self) -> u64 {
+        // TODO get an available unidirectional stream ID more nicely
+        if self.is_server {
+            return 0xB;
+        } else {
+            return 0xA;
+        }
+    }
+
+    pub fn is_established(&mut self) -> bool {
+        self.control_stream_open &&
+            self.qpack_encoder_stream_open &&
+            self.qpack_decoder_stream_open
+    }
+
+    pub fn open_control_stream(&mut self, quic_conn: &mut super::Connection) {
+        if !self.control_stream_open {
+            let mut d = [42; 128];
+            let mut b = octets::Octets::with_slice(&mut d);
```

Comment:
There's no need to use octets here, you can just turn `H3_CONTROL_STREAM_TYPE_ID` into a slice use `to_be_bytes()` method https://doc.rust-lang.org/std/primitive.u8.html#method.to_be_bytes

Reply:
Same below.